docs/topics/extensions.md

```diff
@@ -206,81 +409,198 @@ class Host(val hostname: String) {
+        println("Sending admin notification from special sender")
     }
 }
 
 fun main() {
-    BaseCaller().call(Base())   // "Base extension function in BaseCaller"
-    DerivedCaller().call(Base())  // "Base extension function in DerivedCaller" - dispatch receiver is resolved virtually
-    DerivedCaller().call(Derived())  // "Base extension function in DerivedCaller" - extension receiver is resolved statically
+    // Dispatch receiver is NotificationSender
+    // Extension receiver is User
+    // Resolves to User.sendNotification() in NotificationSender
+    NotificationSender().notify(User())
+    // Sending user notification from normal sender
+    
+    // Dispatch receiver is SpecialNotificationSender
+    // Extension receiver is User
+    // Resolves to User.sendNotification() in SpecialNotificationSender
+    SpecialNotificationSender().notify(User())
+    // Sending user notification from special sender 
+    
+    // Dispatch receiver is SpecialNotificationSender
+    // Extension receiver is User NOT Admin
+    // The notify() function declares user as type User
+    // Statically resolves to User.sendNotification() in SpecialNotificationSender
+    SpecialNotificationSender().notify(Admin())
+    // Sending user notification from special sender 
+}
+```
+{kotlin-runnable="true" kotlin-min-compiler-version="1.3" id="kotlin-extension-open"}
+
+The dispatch receiver is resolved at runtime using virtual dispatch, which makes the behavior in the `main()` function
+easier to follow. What may surprise you is that when you call the `notify()` function on an `Admin` instance, the 
+compiler chooses the extension based on the declared type: `user: User`, because it resolves the extension receiver statically.
+
+## Extensions and visibility modifiers
+
+Extensions use the same [visibility modifiers](visibility-modifiers.md) as regular functions declared in the same scope, including extensions
+declared as members of other classes.
+
+For example, an extension declared at the top level of a file can access other `private` top-level declarations in the same file:
+
+```kotlin
+// File: StringUtils.kt
+
+private fun removeWhitespace(input: String): String {
+    return input.replace("\\s".toRegex(), "")
+}
+
+fun String.cleaned(): String {
+    return removeWhitespace(this)
+}
+
+fun main() {
+    val rawEmail = "  user @example. com  "
+    val cleaned = rawEmail.cleaned()
+    println("Raw:     '$rawEmail'")
+    // Raw:     '  user @example. com  '
+    println("Cleaned: '$cleaned'")
+    // Cleaned: 'user@example.com'
+    println("Looks like an email: ${cleaned.contains("@") && cleaned.contains(".")}") 
+    // Looks like an email: true
 }
 ```
-{kotlin-runnable="true" kotlin-min-compiler-version="1.3"}
+{kotlin-runnable="true" kotlin-min-compiler-version="1.3" id="kotlin-extension-visibility-top-level"}
+
+And if an extension is declared outside its receiver type, it can't access the receiver's `private` or `protected` members:
```

Comment:
Maybe add that `internal` members may only be accessed if you're in the same module?